test/integration/undo-redo/index.spec.ts

```diff
@@ -0,0 +1,82 @@
+import * as path from 'path';
+import { ToolpadEditor } from '../../models/ToolpadEditor';
+import { ToolpadHome } from '../../models/ToolpadHome';
+import { test, expect } from '../../playwright/test';
+import { readJsonFile } from '../../utils/fs';
+import clickCenter from '../../utils/clickCenter';
+
+test('test basic undo and redo', async ({ page, browserName }) => {
+  const homeModel = new ToolpadHome(page);
+  const editorModel = new ToolpadEditor(page, browserName);
+
+  const domInput = await readJsonFile(path.resolve(__dirname, './dom.json'));
+
+  await homeModel.goto();
+  const app = await homeModel.createApplication({ dom: domInput });
+  await editorModel.goto(app.id);
+
+  await editorModel.pageRoot.waitFor();
+
+  const canvasInputLocator = editorModel.appCanvas.locator('input');
+
+  // Initially we should have 2 text fields
+  await expect(canvasInputLocator).toHaveCount(2);
+
+  // Add 3rd text field
+  await editorModel.dragNewComponentToAppCanvas('Text field');
+
+  // Ensure that we added 3rd text field
+  await expect(canvasInputLocator).toHaveCount(3);
+
+  const undoButton = page.locator('[data-testid=undo-button]');
+
+  // Wait for undo stack to be updated
+  await page.waitForTimeout(600);
```

Comment:
Another implementation detail that we should avoid if possible, but again it's hard to think of alternatives in this case...

Reply:
It's not implementation detail. in this case it's a behaviour detail - if you execute actions faster than 500ms they will be batched (something that is technically possible to do with keyboard/mouse)

Reply:
Yeah, I guess it's not to bad, we could maybe pass that timeout + 100 here in case we change the timeout, if possible, other than that seems ok!

Reply:
I assume that after your atomic changes PR is merged this will become obsolete anyways, so we can tweak it together with those changes maybe as throttle will not be needed anymore 🤔

Reply:
I'm keeping the throttling for now as it looks like we'll still need it for undoing or redoing changing text in text inputs for example - so whether we'll keep the throttling or not it depends on our final implementation of that.
Anyway I think using a constant like `UNDO_REDO_THROTTLE = 500 // milliseconds` and then something like `UNDO_REDO_THROTTLE + 100` in the test would be cool but anyway I think this is fine.

Reply:
We shouldn't need throttle in text fields, because when we undo/redo inside text fields, default browser functionality is used 🤔

Reply:
> We shouldn't need throttle in text fields, because when we undo/redo inside text fields, default browser functionality is used

The dom is updated on every keystroke, so that it can reflect immediately in the canvas. The undo stack shouldn't update as often.